Prepare jekyll for Dev blogging

joshmfrankel opened the pull request:
We talked about converting our lessonly.github.io domain into more than just a styleguide so that we could blog as a dev team. This PR is the first step to enabling that. There is now paginated posts and a link in the footer to the current styleguide. The styleguide now resides in a subdirectory.

* Converts styleguide to a sub-directory. Also new styleguide links use the title of their page (css-style-guide) for the text and their filename (css.md) for the url /css
* ~~Adds octopress for cli actions~~ Added jekyll-compose as it is better and more straight-forward
* Adds template blog post (~~/:year~~/:title)
* Removed existing seo and feed logic from <head> tag in favor of seo and feed jekyll plugins
* Removed feed.xml for automatically generated plugin version

## Notes
```
Jekyll-compose
  draft      # Creates a new draft post with the given NAME
  post       # Creates a new post with the given NAME
  publish    # Moves a draft into the _posts directory and sets the date
  unpublish  # Moves a post back into the _drafts directory
  page       # Creates a new page with the given NAME
```

## Testing Notes

* Try the site out locally
- [x] `jekyll serve` and check it out new structure
- [x] Run `jekyll post "Your post name here"` and see how easy it is to create a new blog post
- [ ] /feed.xml still works
- [ ] Proper seo tags in the <head> section

joshmfrankel commented:
@lessonly/developers Love to get everyone's thoughts on this before merging as it changes our styleguide file structure and enables us to start blogging.

stevegrossi commented on Gemfile:
```diff
@@ -1,4 +1,11 @@
 
 gem "jekyll", "~> 3.0"
 gem "github-pages"
+gem "octopress", "~> 3.0"
```
Hm, having trouble installing this:

```
An error occurred while installing json (1.8.3), and Bundler cannot
continue.
Make sure that `gem install json -v '1.8.3'` succeeds before bundling.
steve@lessonly.github.io gem install json -v '1.8.3'
Building native extensions.  This could take a while...
ERROR:  Error installing json:
	ERROR: Failed to build gem native extension.
```

joshmfrankel replied:
Hmm, so I have octopress locally installed so I wonder if I also have json installed. Do you think adding

``` ruby
gem "json", "~>1.8"
gem "octopress", "~> 3.0"
```

in the Gemfile would solve the above issue? We don't really need octopress it really is just great for quickly creating new posts

stevegrossi replied:
I added `gem "json", "~>1.8"` to my Gemfile but still get the error. Hopefully someone else can try this out and confirm if it's just my system.

rreinhardt9 replied:
if we wanted to, we could just drop octopress and user https://github.com/jekyll/jekyll-compose it's just a lightweight jekyll plugin to help with the publishing flow. I've been using that and straight jekyll for my personal blog and have had a good experience.

rreinhardt9 replied:
The nice thing about plain jekyll is that github pages will detect that it's a jekyll site and auto complile it for you... so you don't have to generate the site locally before you push it up. I don't remember how octopress works with that...

stevegrossi replied:
I also love how you can edit posts with like 2 clicks through Github.com, rather that having to do the whole recompile & `git push` dance... If Octopress doesn't support that, put me down for plain jekyll 👍

rreinhardt9 replied:
I was able to close and bundle successfully; I did get this error when trying to build the site:

```
$ bundle exec jekyll serve
Configuration file: /Users/ross/dev/lessonly.github.io/_config.yml
            Source: /Users/ross/dev/lessonly.github.io
       Destination: /Users/ross/dev/lessonly.github.io/_site
 Incremental build: disabled. Enable with --incremental
      Generating... 
             ERROR: YOUR SITE COULD NOT BE BUILT:
                    ------------------------------------
                    Invalid date '<%= Time.now.strftime('%Y-%m-%d %H:%M:%S %z') %>': Document 'vendor/bundle/gems/jekyll-3.0.2/lib/site_template/_posts/0000-00-00-welcome-to-jekyll.markdown.erb' does not have a valid date in the YAML front matter.
```

rreinhardt9 replied:
I think octopress typically has a "source" branch that contains the site (as we currently have it on master) and then you use the octopress rake tasks to "generate" the site onto a master branch that gets pushed to publish the site? Vs. just having plain ol' jekyll on master and letting github parse and publish (that's how we have been doing it)

joshmfrankel replied:
> if we wanted to, we could just drop octopress and use https://github.com/jekyll/jekyll-compose it's just a lightweight jekyll plugin to help with the publishing flow. I've been using that and straight jekyll for my personal blog and have had a good experience.

That looks awesome and I'm good switching to it :+1: 

> The nice thing about plain jekyll is that github pages will detect that it's a jekyll site and auto complile it for you... so you don't have to generate the site locally before you push it up. I don't remember how octopress works with that...

So octopress is just a quick way of running commands like new draft/post/page/etc. You can also do things like octopress deploy with github pages. Though I'm not sure if that really buys us anything here.

I think for the purpose of this initial PR lets go with jekyll-compose as it looks straight-forward

stevegrossi commented on _styleguide/css.md:
```diff
@@ -1,7 +1,6 @@
 ---
 layout: page
 title: CSS Style Guide
-permalink: /css/
```
It's nice not having to specify this! 👍

stevegrossi commented on _posts/2017-06-29-first-post.markdown:
```diff
@@ -0,0 +1,13 @@
+
+{% highlight ruby %}
+Post.create!
+{% endhighlight %}
```
I don't want this to stop us, but it'd be nice to have some actual content for our first post. Maybe a "Lessonly is Hiring Engineers!" post? and we link to www.lessonly.com/hiring/?

joshmfrankel replied:
I actually was having a similar thought here as well. Maybe a quick 200-300 word post about hiring would be perfect

wsmd commented:
> so that we could blog as a dev team

![jim-carey_excited](https://user-images.githubusercontent.com/2100222/27752736-8f29ff00-5db0-11e7-9f98-20d64126edba.gif)

joshmfrankel commented on _posts/2017-06-29-were-hiring-new-engineers.markdown:
```diff
@@ -0,0 +1,11 @@
+---
+
+Want to be part of our product team? We're looking to hire new engineers to join
+the llama herd. Find out more by <a href="http://www.lessonly.com/hiring/">applying here</a>.
```
Probably could use some editing, but here is a brief post to call out that we are hiring. Thanks for the suggestion Steve. :-)

joshmfrankel commented on Gemfile:
```diff
@@ -1,4 +1,11 @@
+  gem "jekyll-paginate", "~> 1.1"
+  gem "jekyll-seo-tag", "~>0.1"
+  gem "jekyll-feed", "~>0.3"
+  gem "jekyll-compose", "~>0.5"
```
@rreinhardt9 jekyll-compose works way better than octopress. Going to have to start using this on my dev blog :-)

rreinhardt9 replied:
:tada: I used to use octopress... but I kinda got tired of some of it's complexity (especially when I started adding my "landing page") so I was going to try and write my own rake tasks for my publishing flow; but then I discovered that someone already had with jeyll-compose so I just started using that :-)

Thanks for pulling this all together by the way! So awesome!

joshmfrankel commented:
@stevegrossi I removed octopress in favor of jekyll-compose. Hopefully that allows you to look at this locally. Thanks for the suggestion @rreinhardt9. Also I added a basic post to call out that we are hiring.

stevegrossi commented:
I still get the same JSON gem error, but it looks like a known issue with Ruby 2.4 (my default) and the version of `json` our Rails is pegged to: https://github.com/rails/rails/issues/27450 I did a `bundle update json` for our jekyll site so it gets the newer version that works with Ruby 2.4 and it's working fine 👍 

We should probably specify a `.ruby-version` in this project so we don't run into Ruby/gem conflicts like this in the future but that can wait.

stevegrossi commented on _styleguide/release-notes.md:
```diff
@@ -1,7 +1,6 @@
 ---
 layout: page
 title: Release Notes
-permalink: /release-notes/
```
Our Release Notes aren't actually part of the styleguide. Can we keep those in the root of the project `/release-notes.md` and hosted at `/release-notes/` instead of `/styleguide/release-notes/`?

joshmfrankel commented:
Alright here's what I added:

* Release notes are now back within /release-notes
* Release notes have been removed from the /styleguide listing page and
moved into the footer navigation
* Posts are categorized by /:title/ only now for smaller and more readable
urls
* Updated the readme with notes on how to add new styleguide content and
writing new blog posts
* Updated date parsing for index listing to match post layout

joshmfrankel commented on _includes/footer.html:
```diff
@@ -4,6 +4,9 @@
     <p>
       Made with love by <a href="http://www.lessonly.com/">Lessonly</a>.
+      | <a href="{{ site.github.url }}/styleguide">Styleguide</a>
+      | <a href="{{ site.github.url }}/release-notes">Release Notes</a>
```
Release notes now are linked in the footer

joshmfrankel commented on feed.xml:
```diff
@@ -1,30 +0,0 @@
-      </item>
-    {% endfor %}
-  </channel>
-</rss>
```
Unnecessary as jekyll-feed does this automatically

stevegrossi commented on .ruby-version:
```diff
@@ -0,0 +1 @@
+2.4
```
Which ruby version manager do you use? I use asdf and it doesn't recognize "2.4", only the full released versions like `2.4.0` and `2.4.1`:

```
steve@lessonly.github.io asdf list-all ruby | grep 2.4
2.4.0-preview1
2.4.0-preview2
2.4.0-preview3
2.4.0-rc1
2.4.0
2.4.1
```

joshmfrankel replied:
Whoops, you are totally right. Forgot to re-run bundle this won't work for me either. Let's make it 2.4.1

joshmfrankel commented:
@stevegrossi Now relies on 💎 2.4.1 in the .ruby-version file

stevegrossi commented:
Looks great locally! Thanks for doing this 💯
